keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md

```diff
@@ -207,6 +245,49 @@ Since falling back to etcd won't work, we should fail the requests and rely on
 rate limiting to prevent cascading failure.  I.e. `Retry-After` HTTP header (for
 well behaved clients) and [Priority and Fairness](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md).
 
+For such situations we will provide users with following tools:
+* a dedicated `apiserver_watch_cache_read_wait` metric to detect a problem with
+  watch cache.
+* a per-request override via `X-DEBUG: NO-WATCH-CACHE` HTTP header for
+  debugging.
+* an escape hatch `--apiserver-consistent-list-from-cache` flag to disable the
+  feature.
+
+Metric `apiserver_watch_cache_read_wait` will measure wait time experienced by 
+reads for watch cache to become fresh. If user notices a latency request in
+they can use this metric to confirm that the issue is caused by watch cache.
+
+Per request override via an HTTP request header `X-DEBUG: NO-WATCH-CACHE`, 
+should allow user to compare request results without impacting other requests or
+requiring to redeploy whole cluster. We propose to use a header, as we don't
+want to create an official API as we don't expect those requests to be ever be 
+executed for other purposed than debugging. In healthy situation setting this 
+header should not cause any impact on the resulting response, however it might
+have performance impact, by increasing load on etcd. To prevent abuse we will
+limit access to users with `cluster-admin` role, rejecting the request otherwise.
+
+To execute a request with watch cache disabled:
+
+```shell
+kubectl proxy &
+curl -H 'X-DEBUG: NO-WATCH-CACHE' http://localhost:8080/api/v1/namespaces/my-ns/pods/my-pod?resourceVersion=0
+```
+
+A `--apiserver-consistent-list-from-cache` flag to disable usage of watch cache
```

Comment:
We have flags to disable the entire watch cache today.  I understand "what if there is a bug" concern, but rather than grow a permanent flag, I'd like to see us develop confidence and once we have it, remove the featuregate.

If this is born out of a concern for storage backends that might not support all etcd features, I think this is exactly the reason why apimachinery took the position of "etcd is the only backend, you can replace it by replacing the etcd functionality".  If a backend doesn't support the etcd contract, they will need to update to handle it.

Reply:
Makes sense.